app/src/main/java/com/concordium/wallet/data/preferences/Preferences.kt

```diff
@@ -15,9 +19,145 @@ open class Preferences {
     interface Listener {
         fun onChange()
     }
+    companion object{
+        const val SHARED_PREFERENCES_ARE_ENCRYPTED ="encrypted_shared_preference"
+    }
 
     constructor(context: Context, preferenceName: String, preferenceMode: Int) {
-        sharedPreferences = context.getSharedPreferences(preferenceName, preferenceMode)
+        val authPreferences = initializeEncryptedSharedPreferences(context, SharedPreferencesKeys.PREF_FILE_AUTH)
+        if(authPreferences.getBoolean(SHARED_PREFERENCES_ARE_ENCRYPTED, false)){
+            sharedPreferences = initializeEncryptedSharedPreferences(context, preferenceName)
+        }else{
+            sharedPreferences =
+                if(migratePreferencesSuccess(context, preferenceName, preferenceMode, authPreferences)){
+                    initializeEncryptedSharedPreferences(context, preferenceName)
+                }else{
+                    getSharedPreferences(context, preferenceName, preferenceMode)
+                }
+        }
+    }
+
+    /**
+     * Returns the required SharedPreferences
+     * @return [SharedPreferences] *if* unencrypted data is still present
+     *
+     * *else* an instance of encrypted [SharedPreferences]
+     */
+    private fun getSharedPreferences(
+        context: Context,
+        preferenceName: String,
+        preferenceMode: Int
+    ): SharedPreferences {
+        val unEncryptedSharedPreferences = context.getSharedPreferences(preferenceName, preferenceMode)
+
+        return if(unEncryptedSharedPreferences.all.isEmpty()) {
+            initializeEncryptedSharedPreferences(context, preferenceName)
+        }else{
+            unEncryptedSharedPreferences
+        }
+    }
+
+    /**
+     * Loops through all of the [SharedPreferences] and attempts to save them to [EncryptedSharedPreferences]
+     * @return *true* if the [SharedPreferences] don't need migrating
+     *
+     * *true* if [SharedPreferences] is successfully migrated and the old data cleared
+     *
+     * *false* if the migrations fail or is not cleared successfully
+     */
+    private fun migratePreferencesSuccess(
+        mContext: Context,
+        preferenceName: String,
+        preferenceMode: Int,
+        authPreferences: SharedPreferences): Boolean{
+        var allPreferencesAreMigrated = true
+        var continueWithEncryptedSharedPreference = true
+
+        for(prefName in SharedPreferencesKeys.PREF_ALL){
+            if(!migrateSinglePreferencesIfNeededOrContinue(mContext, prefName, preferenceMode)){
+                allPreferencesAreMigrated = false
+
+                if(prefName == preferenceName) {
+                    continueWithEncryptedSharedPreference = false
+                }
```

Comment:
What is the purpose of this check? It's confusing to me that there is both a `preferenceName` and `prefName`. The variable names should indicate what the difference between them is.

Reply:
We are looping through all of the SharedPreferences names and trying to migrate them, the current item is `prefName` and our working preference is `preferenceName`. So if the migrations fails for a SharedPreference we check if that is our working preference, if it is then we tell it to not use the encrypted shared preference.